src/cc2olx/tools/video_download.py

```diff
@@ -0,0 +1,181 @@
+import argparse
+import json
+import csv
+import zipfile
+from urllib.parse import parse_qs, urlparse
+
+from lxml import html
+
+import youtube_dl
+
+
+def parse_args():
+    """Parse command line arguments."""
+
+    parser = argparse.ArgumentParser(
+        description="Find and download videos from a CC course."
+    )
+    parser.add_argument("--config", "-c", help="JSON config file")
+    parser.add_argument("--input", "-i", help="Input CC zip or HTML file")
+    parser.add_argument("--output", "-o", default="out.csv", help="CSV output file")
+    parser.add_argument(
+        "--downloads", "-d", default="downloads", help="Video download directory"
+    )
+    parser.add_argument(
+        "--simulate", "-s", action="store_true", help="Simulate downloads"
+    )
+    parser.add_argument(
+        "--verbose", "-v", action="store_true", help="Verbose ytd downloads"
+    )
+
+    return parser.parse_args()
+
+
+def get_ydl_opts(args):
+    """Set options to be used by Youtube DL"""
+
+    if args.config:
+        return json.load(open(args.config))
+
+    return {
+        "verbose": args.verbose,
+        "simulate": args.simulate,
+        "writesubtitles": True,
+        "writeautomaticsub": True,
+        "outtmpl": f"{args.downloads}/%(title)s.%(ext)s",
+    }
+
+
+def download_videos(urls, opts):
+    """
+    Download a list of videos and optionally transcripts.
+    Arguments:
+        * videos: a list of URLs
+        * opts: options for Youtube DL
+    """
+
+    max = len(urls)
+    #max = 3 #stop after a few video downloads, for debugging only
+
+    urls = urls[:max]
+    relpaths = []
+
+    def my_hook(d):
+        """Youtube DL callback"""
+        if d['status'] == 'finished':
+            fn = d['filename']
+            if not fn.endswith('m4a'):
+                relpaths.append(fn)
+
+    opts['progress_hooks'] = [my_hook]
+
+    for url in urls:
+        print('Downloading', url)
+        with youtube_dl.YoutubeDL(opts) as ydl:
+            ydl.download([url])
+
+    return relpaths
+
+
+def make_row(relpath, url):
+    """Create an entry for the output CSV."""
+
+    youtube_id = ""
+    if "youtube.com" in url:
+        youtube_id = url.split("=")[-1]
+
+    return {
+        "Relative File Path": relpath,
+        "External Video Link": url,
```

Comment:
This writes the reformatted URLs to the CSV, while [the iframe_link_parser uses this field to search the iframe sources for the original URL](https://github.com/edx/cc2olx/blob/bff10a2ed6be7f8701e29ed81275a00530b134a0/src/cc2olx/iframe_link_parser.py#L80).
I think we'll need to leave these as the original URL we find in the iframe for cc2olx to properly locate them to convert them to the edX ID.

Reply:
_get_video_url() in the link parser returns a formatted url and uses that as a key in the link map.


Right now cc2olx does not work with the CSV this script produces, because it needs an upload script to add the edx_video_id columns. And it's not easy to test without uncommenting max = 3 or finding a CC course that has less the 600 videos to download. I'll find a way to make testing only the generation of the link map easier.

Reply:
Ah my mistake, then this should be fine.

> Right now cc2olx does not work with the CSV this script produces, because it needs an upload script to add the edx_video_id columns.

Right, that's expected that `video_upload.py` will be used before running this through cc2olx. Yes I Had some trouble testing as well and even ran out of hard drive space while trying to run `video_download.py` over an entire course.